### PR TITLE
remoteproc_virtio: optimize the remoteproc virtio transport layer

### PR DESCRIPTION
**Pull request by CV-Bowen**
remoteproc_virtio: optimize the remoteproc virtio transport layer

1. Implement the rproc_virtio_create_virtqueues() and rproc_virtio_delete_virtqueues()
2. Becase item 1, modified the `rpmsg_init_vdev_with_config()` also

### CODE REVIEW

**Comment by arnopo** [comment]
hello @CV-Bowen ,

Could you details what was the issue you are facing that needs to move the virtio_create_virtqueues to the transport layer?
that would help to understand your work

**Comment by CV-Bowen** [comment]
@arnopo Hi, we want to support different transport layer (remoteproc, mmio) for virtio drivers. After this PR, we can implement the mmio tranport layer base on OpenAmp.
I think this patch is also helpful for virtio-mmio framework in openamp/virtio-exp branch to support remoteproc transport layer.

File: lib/include/openamp/virtio.h
Comment (edmooring) [inline lib/include/openamp/virtio.h:84]: I don't see this field being used anywhere. Why is it needed?

Comment (CV-Bowen) [reply lib/include/openamp/virtio.h]: I delete this, thanks.

**Comment by arnopo** [comment]
> @arnopo Hi, we want to support different transport layer (remoteproc, mmio) for virtio drivers. After this PR, we can implement the mmio tranport layer base on OpenAmp. I think this patch is also helpful for virtio-mmio framework in openamp/virtio-exp branch to support remoteproc transport layer.

I understand the objective., what is not clear to me as a first review is why you need to move `virtio_create_virtqueues` this should be independent from the transport layer, no?

File: lib/remoteproc/remoteproc_virtio.c
Comment (arnopo) [inline lib/remoteproc/remoteproc_virtio.c:59]: I can not see something related to remoteproc virtio transport in function above, that why  I need to understand more in details what would be the delta with a virtio MMIO

**Review by arnopo — CHANGES_REQUESTED** [review]

**Comment by CV-Bowen** [comment]
> > @arnopo Hi, we want to support different transport layer (remoteproc, mmio) for virtio drivers. After this PR, we can implement the mmio tranport layer base on OpenAmp. I think this patch is also helpful for virtio-mmio framework in openamp/virtio-exp branch to support remoteproc transport layer.
> 
> I understand the objective., what is not clear to me as a first review is why you need to move `virtio_create_virtqueues` this should be independent from the transport layer, no?

I found only `rpmsg_virtio_create_virtqueues` call `virtio_create_virtqueues` in OpenAmp, so I delete `virtio_create_virtqueues` after moving it's implementation to `rproc_virtio_create_virtqueues`. I do not consider that some other codes use it, I will restore this api and call it directly in `rproc_virtio_create_virtqueues()`.

File: lib/include/openamp/virtio.h
Comment (arnopo) [reply lib/include/openamp/virtio.h:168]: > But I see the virtqueues create and delete in linux are in transport layer. See https://github.com/torvalds/linux/blob/master/drivers/virtio/virtio_mmio.c#L490 and https://github.com/torvalds/linux/blob/master/drivers/virtio/virtio_mmio.c#L354

Ok i start to understand your need... 

The equivalent of the `virtio_dispatch` is the [`virtio_config_ops`](https://elixir.bootlin.com/linux/latest/source/include/linux/virtio_config.h#L106)

so you need an equivalent of `find_vqs` and `del_vqs`

Now I would like to understand why the backends can not call `virtio_create_virtqueues` instead of duplicating the code ?

Comment (CV-Bowen) [reply lib/include/openamp/virtio.h:168]: I'm busy with other projects recently and very sorry for the late reply.

> The equivalent of the virtio_dispatch is the [virtio_config_ops]
> so you need an equivalent of find_vqs and del_vqs

Yes, I reference the Linux design.

> Now I would like to understand why the backends can not call virtio_create_virtqueues instead of duplicating the code ?

Because different transport layer has different implementation, e.g. mmio transport layer need set the virtqueue address but the remoteproc do not need.

File: lib/include/openamp/virtio.h
Comment (CV-Bowen) [reply lib/include/openamp/virtio.h:169]: OK, but after I rename the "virtio_device" to "virtio", the function names are duplicated. Maybe rename the `create_virtqueues` in `struct virtio_dispatch` to other name is better?

File: lib/include/openamp/virtio.h
Comment (CV-Bowen) [reply lib/include/openamp/virtio.h:209]: Seems change to macro is better.

**Comment by CV-Bowen** [comment]
@arnopo @edmooring
Commit 1: virtio: follow virtio 1.2 spec, add more virtio status and device id
Seems commit 1 can be an independent PR. I will upload it later, and this PR will be only for Commit 2.

File: lib/include/openamp/virtio.h
Comment (arnopo) [reply lib/include/openamp/virtio.h:168]: > I'm busy with other projects recently and very sorry for the late reply.

No worries, It is thesame for everyone ;-)

> > Now I would like to understand why the backends can not call virtio_create_virtqueues instead of duplicating the code ?
> 
> Because different transport layer has different implementation, e.g. mmio transport layer need set the virtqueue address but the remoteproc do not need.

If you have some code to share, I'm interested in. That would help me to understand your point

File: lib/include/openamp/rpmsg_virtio.h
Comment (arnopo) [inline lib/include/openamp/rpmsg_virtio.h:145]: In fact doing that you break some project compatibility:
for instance
https://elixir.bootlin.com/zephyr/latest/source/subsys/ipc/rpmsg_service/rpmsg_backend.c#L145

At least need to support legacy using deprecating method.

File: lib/include/openamp/virtio.h
Comment (arnopo) [reply lib/include/openamp/virtio.h:169]: > OK, but after I rename the "virtio_device" to "virtio", the function names are duplicated. Maybe rename the `create_virtqueues` in `struct virtio_dispatch` to other name is better?

Right, or perhaps adding `virtio_dispatch_xxxx` or `virtio_trasport_xxxx` API for the dispatch ops.

Comment (CV-Bowen) [reply lib/include/openamp/virtio.h]: I rename them to virtio_dispatch_xxxx  and change them to macro.

File: lib/include/openamp/virtio.h
Comment (CV-Bowen) [reply lib/include/openamp/virtio.h]: Done

File: lib/include/openamp/virtio.h
Comment (CV-Bowen) [reply lib/include/openamp/virtio.h]: Done

**Comment by arnopo** [comment]
@CV-Bowen 
seems that you remove the Commit 2: virtio: decoupling the transport layer and virtio device layer

Please notice also https://github.com/OpenAMP/open-amp/pull/494, we should find a common solution...

**Comment by CV-Bowen** [comment]
@arnopo Sorry, I has fixed this and I will look #494.

File: lib/include/openamp/rpmsg_virtio.h
Comment (xiaoxiang781216) [reply lib/include/openamp/rpmsg_virtio.h:145]: let's check create_virtqueues and do the different dispatch

File: lib/include/openamp/virtio.h
Comment (xiaoxiang781216) [reply lib/include/openamp/virtio.h:168]: @arnopo I think you get the sample code from here:
https://github.com/OpenAMP/open-amp/pull/494/commits/ae651f3c9970fe12dd00c16c25eff5e5b64596e8#r1227687251
@CV-Bowen just extract the core change from @danmilea and make it generalize to support both remoteproc and mmio(or pci) transport backend.

File: lib/remoteproc/remoteproc_virtio.c
Comment (xiaoxiang781216) [reply lib/remoteproc/remoteproc_virtio.c:151]: Please reference this pr: https://github.com/OpenAMP/open-amp/pull/494

**Comment by xiaoxiang781216** [comment]
> @CV-Bowen seems that you remove the Commit 2: virtio: decoupling the transport layer and virtio device layer
> 
> Please notice also #494, we should find a common solution...

@CV-Bowen 's change extract and generalize from #494, after this change get merged:

1. The new mmio transport addition don't need modify rest code in OpenAMP
2. All code manipulate virtio device can work with both transport(mmio v.s. remoteproc)

File: lib/remoteproc/remoteproc_virtio.c
Comment (arnopo) [inline lib/remoteproc/remoteproc_virtio.c:45]: unexpected dependency with remoteproc.c ( this structure does not exist for some projects  implementing device part)

File: lib/remoteproc/remoteproc_virtio.c
Comment (arnopo) [inline lib/remoteproc/remoteproc_virtio.c:59]: priv can be something else than rproc

File: lib/remoteproc/remoteproc_virtio.c
Comment (arnopo) [inline lib/remoteproc/remoteproc_virtio.c:71]: this can probably be kept in remoteproc_create_virtio

File: lib/rpmsg/rpmsg_virtio.c
Comment (arnopo) [inline lib/rpmsg/rpmsg_virtio.c:871]: to put under `#ifndef VIRTIO_DEVICE_ONLY` to fix build warning

File: lib/include/openamp/rpmsg_virtio.h
Comment (arnopo) [reply lib/include/openamp/rpmsg_virtio.h:145]: yes something like 
```
if (rvdev->vdev->func->create_virtqueues) {
	return rvdev->vdev->func->create_virtqueues(rvdev->vdev, flags, nvqs,
						    names, callbacks);
} else {
	metal_log(METAL_LOG_WARNING,
				  "deprecated method please set create_virtqueues instead");
	virtio_create_virtqueues(rvdev->vdev, flags, nvqs, names,
					callbacks);
}
```

**Review by arnopo — CHANGES_REQUESTED** [review]
This PR probably breaks a lot of implementations. 
I haven't done a test, but I'm fairly convinced that it breaks all Zephyr and ST projects, for example.

One reason is that you introduce a dependency between `remoteproc` and `remoteproc_virtio`, using `struct remoteproc`
Another is that some projects don't use the remoteproc virtio at all but static vrings.

Could  you also split the commit in 3 ( as described in your commit message) ?
This would help for the review...

**Comment by arnopo** [comment]
@CV-Bowen :
Please, could you push your work on lib/include/openamp/virtio.h in a separate pull request as requested in (https://github.com/OpenAMP/open-amp/issues/390#issuecomment-1602605560) 
That well help to move forward on the topic?
Thanks in advance.

**Comment by CV-Bowen** [comment]
@arnopo Sorry, i'm busy with other things, and the new PR has been created ( https://github.com/OpenAMP/open-amp/pull/495). But a mistake is this PR is also updated, i will restore this PR.

File: lib/include/openamp/virtio.h
Comment (arnopo) [inline lib/include/openamp/virtio.h:226]: Please use inline function instead. `checkpatch --strict` complains on this and followings macro.
`CHECK:MACRO_ARG_REUSE: Macro argument reuse 'vdev' - possible side-effects?`

**Review by arnopo — APPROVED** [review]
Thank you for your prompt update!

LGTM after addressing the 2 comments and fix "deivce" typo in commit

**Comment by arnopo** [comment]
@CV-Bowen 
Zephyr Build fails because the sample application has defined same functions(https://elixir.bootlin.com/zephyr/latest/source/samples/subsys/ipc/openamp/src/main.c#L66).
 Just ignore the issue this has to be fixed in Zephyr ( I will send a PR)

**Comment by arnopo** [comment]
> Thank you for your prompt update!
> 
> LGTM after addressing the 2 comments and fix "deivce" typo in commit

Just see that you create the https://github.com/OpenAMP/open-amp/pull/495 to create the dispatch functions.
so Please take into account my remarks to apply them on https://github.com/OpenAMP/open-amp/pull/495.

**Comment by CV-Bowen** [comment]
@arnopo  Done in #495

**Comment by CV-Bowen** [comment]
I will continue this PR to implement the remoteproc transport ops: create_virtqueue and delete_virtqueues.

File: lib/include/openamp/rpmsg_virtio.h
Comment (CV-Bowen) [reply lib/include/openamp/rpmsg_virtio.h]: This has been implmented in other PR: https://github.com/OpenAMP/open-amp/pull/495

File: lib/virtio/virtio.c
Comment (CV-Bowen) [reply lib/virtio/virtio.c]: Done

File: lib/remoteproc/remoteproc_virtio.c
Comment (CV-Bowen) [reply lib/remoteproc/remoteproc_virtio.c]: Done

File: lib/remoteproc/remoteproc_virtio.c
Comment (CV-Bowen) [reply lib/remoteproc/remoteproc_virtio.c]: Done

File: lib/remoteproc/remoteproc_virtio.c
Comment (CV-Bowen) [reply lib/remoteproc/remoteproc_virtio.c]: Nice suggestions and Done.

File: lib/include/openamp/rpmsg_virtio.h
Comment (CV-Bowen) [reply lib/include/openamp/rpmsg_virtio.h]: Done, This has been implmented in other PR: https://github.com/OpenAMP/open-amp/pull/495

File: lib/rpmsg/rpmsg_virtio.c
Comment (CV-Bowen) [reply lib/rpmsg/rpmsg_virtio.c:865]: Thanks, done

File: lib/include/openamp/virtio.h
Comment (CV-Bowen) [reply lib/include/openamp/virtio.h]: Done

File: lib/virtio/virtio.c
Comment (CV-Bowen) [reply lib/virtio/virtio.c]: Done, This has been implmented in other PR: https://github.com/OpenAMP/open-amp/pull/495

**Comment by CV-Bowen** [comment]
All the comments should have been addressed. Could you take a look again? @arnopo

File: lib/remoteproc/remoteproc_virtio.c
Comment (CV-Bowen) [reply lib/remoteproc/remoteproc_virtio.c]: Yes, this is a mistake, thanks and done.

Comment (xiaoxiang781216) [reply lib/remoteproc/remoteproc_virtio.c]: LGTM.

File: lib/include/openamp/rpmsg_virtio.h
Comment (arnopo) [inline lib/include/openamp/rpmsg_virtio.h:171]: virtio_delete_virtqueues return a value that is just ignored here.

File: lib/rpmsg/rpmsg_virtio.c
Comment (arnopo) [inline lib/rpmsg/rpmsg_virtio.c:888]: `virtio_delete_virtqueues(struct virtio_device *vdev)` returns -ENXIO if `vdev->func->delete_virtqueues` not defined.
But delete_virtqueues is optional.

To be compatible, seems that the virtio_delete_virtqueues should be reworked  to suppress the return .
I pushed a PR in this way to fix this before the release

https://github.com/OpenAMP/open-amp/pull/527

could you review it Please

File: lib/remoteproc/remoteproc_virtio.c
Comment (arnopo) [inline lib/remoteproc/remoteproc_virtio.c:88]: I would prefer that you return an error

File: lib/remoteproc/remoteproc_virtio.c
Comment (arnopo) [inline lib/remoteproc/remoteproc_virtio.c:93]: squash on 1 line

File: lib/remoteproc/remoteproc_virtio.c
Comment (arnopo) [inline lib/remoteproc/remoteproc_virtio.c:94]: useless brackets

File: lib/remoteproc/remoteproc_virtio.c
Comment (arnopo) [inline lib/remoteproc/remoteproc_virtio.c:59]: squash on 1 line

File: lib/remoteproc/remoteproc_virtio.c
Comment (arnopo) [inline lib/remoteproc/remoteproc_virtio.c:52]: useless brackets

File: lib/remoteproc/remoteproc_virtio.c
Comment (arnopo) [inline lib/remoteproc/remoteproc_virtio.c:62]: squash on 1 line and add a blank line between after variable declaration

File: lib/remoteproc/remoteproc_virtio.c
Comment (arnopo) [inline lib/remoteproc/remoteproc_virtio.c:69]: useless brackets + 1 blank line before the return

File: lib/remoteproc/remoteproc_virtio.c
Comment (arnopo) [inline lib/remoteproc/remoteproc_virtio.c:83]: to check that vdev is not null

**Review by arnopo — CHANGES_REQUESTED** [review]
This PR seems Ok in its design but will require some tests to ensure that it does not introduce regression in legacy.
Please find some comments on code

File: lib/include/openamp/rpmsg_virtio.h
Comment (CV-Bowen) [reply lib/include/openamp/rpmsg_virtio.h:171]: Has done in https://github.com/OpenAMP/open-amp/pull/527, let's wait this PR merged and rebase to last.

File: lib/remoteproc/remoteproc_virtio.c
Comment (CV-Bowen) [reply lib/remoteproc/remoteproc_virtio.c:63]: Done

File: lib/remoteproc/remoteproc_virtio.c
Comment (CV-Bowen) [reply lib/remoteproc/remoteproc_virtio.c:84]: vdev has been checked in `virtio_create_virtqueues()`

File: lib/rpmsg/rpmsg_virtio.c
Comment (CV-Bowen) [reply lib/rpmsg/rpmsg_virtio.c:888]: Thanks, LGTM

File: lib/rpmsg/rpmsg_virtio.c
Comment (CV-Bowen) [reply lib/rpmsg/rpmsg_virtio.c]: Done

File: lib/remoteproc/remoteproc_virtio.c
Comment (CV-Bowen) [reply lib/remoteproc/remoteproc_virtio.c]: vdev has been checked in virtio_delete_virtqueues()

File: lib/remoteproc/remoteproc_virtio.c
Comment (CV-Bowen) [reply lib/remoteproc/remoteproc_virtio.c]: Done

File: lib/remoteproc/remoteproc_virtio.c
Comment (CV-Bowen) [reply lib/remoteproc/remoteproc_virtio.c]: Done

File: lib/remoteproc/remoteproc_virtio.c
Comment (CV-Bowen) [reply lib/remoteproc/remoteproc_virtio.c]: Done

File: lib/remoteproc/remoteproc_virtio.c
Comment (CV-Bowen) [reply lib/remoteproc/remoteproc_virtio.c]: Done

File: lib/remoteproc/remoteproc_virtio.c
Comment (CV-Bowen) [reply lib/remoteproc/remoteproc_virtio.c]: Done

File: lib/include/openamp/rpmsg_virtio.h
Comment (CV-Bowen) [reply lib/include/openamp/rpmsg_virtio.h:229]: Done

File: lib/remoteproc/remoteproc_virtio.c
Comment (arnopo) [reply lib/remoteproc/remoteproc_virtio.c:84]: test done in in `virtio_delete_virtqueues()`

**Review by arnopo — APPROVED** [review]
LGTM, I approve it but will need few test before merging it

File: lib/remoteproc/remoteproc_virtio.c
Comment (edmooring) [inline lib/remoteproc/remoteproc_virtio.c:49]: Minor: This comment is misleading. The code would be clearer if it was left out. A more accurate comment would be something like:
"Fail if the virtqueue has already been created"

**Review by edmooring — APPROVED** [review]
Looks good to go, but agree with @arnopo we need to test more. It would be nice if my comment about the comment was addressed, but it is not necessary.

File: lib/remoteproc/remoteproc_virtio.c
Comment (CV-Bowen) [reply lib/remoteproc/remoteproc_virtio.c]: Addressed, thanks, looks better now.

**Comment by tnmysh** [comment]
@arnopo, @edmooring If you are okay, we can merge the change.

**Comment by arnopo** [comment]
@CV-Bowen,
Please, could you rebase it to solve merge conflict, that i test it and merge it

**Comment by CV-Bowen** [comment]
@arnopo Rebased.

**Comment by arnopo** [comment]
tests Ok on my side
@edmooring , @tnmysh , could you make a short test on Xilinx platform for complementary  test on remoteproc ?

**Comment by CV-Bowen** [comment]
@arnopo Rebase to the last to fix the conflict. Could we merge this PR? There isn’t much time until release in April.

**Comment by arnopo** [comment]
Sorry @CV-Bowen, it go out of my radar,

@edmooring , @tnmysh , as requested could you confirm that it does not vcreate regression on your platforms

**Review by tnmysh — APPROVED** [review]

**Pull request merged**
